chore: Standardise native public input handling

federicobarbacovi opened the pull request:
Public inputs in `stdlib` are handled via the special public input mechanisms: we have classes (`KernelIO`, `DefaultIO`, `HidingKernelIO`, `RollupIO`) that set public inputs, reconstruct them, create defaults, etc. Native public inputs have a similar mechanism, but the mechanism has many hacks built-in as native structures are used in more extensively than their `stdlib` counterparts (e.g., we use `field` also to construct the base/scalar field of `secp256k1`, which we never need to deserialise from public inputs).

This PR standardises usage of native public inputs by creating a structure similar to the one used in `stdlib`: we define a `PublicInputComponent` class that works as a wrapper around classes that can be deserialised from the public inputs, and we use this wrapper class to reconstruct special public inputs (pairing points, IPA claims, etc.).

To use the class, we make choices about how elements that are not currently used in circuit should be serialised. This pertains to base/scalar fields of `secp256k1` and `secp256r1` (and therefore to affine points on these curves). For consistency with the base field of `BN254`, we set `PUBLIC_INPUTS_SIZE = 4` for an element in each of these fields.

This PR also cleans up the usage of the constants `PAIRING_POINTS_SIZE` and `IPA_CLAIM_SIZE`, as they can now be extracted from the respective classes `PairingPoints` and `OpeningClaim<Curve>` where `Curve` is `Grumpkin` (either native or `stdlib`)

Closes https://github.com/AztecProtocol/barretenberg/issues/1478

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

suyash67 committed:
chore: remove `logic` operations from `uint` (#15975)

In `stdlib_uint` we no longer need logical operations because the only
places they were used in, i.e., std/turbo version of sha256, blake2s,
blake3s, have been removed. So its best to reduce complexity of the
`uint` class and keep it minimal.

Removed the following functions from the `uint` class:
```cpp
operator^
operator&
operator|
operator~
operator>>
operator<<
ror
rol
logic_operator
```

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
chore: Modify the `MergeVerifier` so that it gets the subtable commitments as input and returns the commitment to the merged table (#15949)

We modify the `MergeVerifier` so that it gets the subtable commitments as input and returns the commitment to the merged table. The reason for this change is that given the new structure of `ClientIVC` following [#15704](https://github.com/AztecProtocol/aztec-packages/pull/15704), we can't access the merged table commitments from inside `complete_hiding_circuit_logic`.

This PR is in preparation for [#15829](https://github.com/AztecProtocol/aztec-packages/pull/15829)

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

johnathan79717 committed:
fix: delete tar generated by test-vk-havent-changed script (#15988)

The tar is not supposed to be checked in according to
https://github.com/AztecProtocol/aztec-packages/pull/15953/commits/50a1bd3e24e412bcf1f4a374a2d81519a7331606#r2229436867

Commit:
Merge branch 'next' into merge-train/barretenberg

johnathan79717 committed:
chore: replace q_arith with q_3 in memory relation (#15953)

See https://github.com/AztecProtocol/barretenberg/issues/914

Commit:
Merge branch 'next' into merge-train/barretenberg

suyash67 committed:
chore: simplify `uint` logic by removing `witness_status` (#15976)

TLDR: `uint` arithmetic operators `+` and `-` had a coding error and as
a result, we weren't actually supporting lazy arithmetic over integers.
This PR simplifies the `uint` class to now allow any "unbounded" values.

#### The Issue 

In the current `uint` class, we allow "unbounded" values, for example, a
`uint32_ct` can contain a value > 32 bits. This was done to allow lazy
arithmetic before such values were "normalized". This is because a call
to `normalize()` is expensive: it decomposes the value in 12-bit slices
and range-constrains each slice.

In practice though, the addition and subtraction operator actually
didn't allow any overflow due to a coding error.
On adding two $\textsf{uint}x$ values $a$ and $b$ (where $x \in [8, 16,
32, 64]$), we currently do:


https://github.com/AztecProtocol/aztec-packages/blob/5c2c217a2f1b05ae226a16ee19a99079dbba8fec/barretenberg/cpp/src/barretenberg/stdlib/primitives/uint/arithmetic.cpp#L27-L47

Assume $a, b$ are both witnesses, the `create_balanced_add_gate` creates
the following constraint:

$$a + b = q \cdot \textcolor{grey}{2^x} + r$$

where the quotient $q$ and remainder $r$ are computed as:

$$q := \frac{(a \textsf{ mod } 2^x) + (b \textsf{ mod } 2^x)}{2^x},
\quad r := \left((a \textsf{ mod } 2^x) + (b \textsf{ mod } 2^x)\right)
\textsf{ mod } 2^x.$$

In other words, the quotient and remainder are computed from the
"truncated" values of $a$ and $b$ when it should have been from the
"unbounded" values. Effectively, this means we are not actually
supporting lazy arithmetic (i.e., arithmetic operations expect inputs to
be "normalized"). I wrote a test
[here](https://github.com/AztecProtocol/aztec-packages/blob/ace0afdb4fb773cfc50af92930ecb94993ab72a5/barretenberg/cpp/src/barretenberg/stdlib/primitives/uint/uint.test.cpp#L243-L271)
that fails when, ideally, it should have passed. This confirmed the
coding error.

#### Solution(s)

One way to fix this is to actually use `get_unbounded_value()` in place
of `get_value()` (on lines 27 and 28 in `operator+` above). But we never
really were using the benefits of lazy addition (because of this silly
error). So we decided its better to remove functionality related to
"unbounded" uint values.

Thus, we remove the `witness_status` member of the `uint` class as it
tracks if a `uint` needs to be "normalized". As a consequence, we now
need to "normalize" in every constructor where we weren't constraining
the accumulators (i.e., `byte_array` and `std::vector<bool_t>`).
Further, in `operator+` and `operator-` we normalize the result. Also,
removed the `get_unbounded_value()` as it isn't being used anywhere.

suyash67 committed:
Revert "chore: remove `logic` operations from `uint`" (#15997)

Reverts AztecProtocol/aztec-packages#15975

Need to get the logic ops cleanup
[PR](https://github.com/AztecProtocol/aztec-packages/pull/15823) first.

suyash67 committed:
Revert "chore: simplify `uint` logic by removing `witness_status`" (#16000)

Reverts AztecProtocol/aztec-packages#15976

federicobarbacovi committed:
feat: Add the last merged table to the public inputs of the hiding circuit (#15829)

We make the merged table received by the Merge verifier in the hiding
circuit a public input to the hiding circuit. This is needed because the
Merge verifier will soon receive `t_commitments`, `T_prev_commitments`
as inputs rather than reading them from the proof.

**EDIT:**

To complete the work on the consistency checks, and to ensure the
soundness of the Goblin verification, the merged table received by the
Merge verifier in the last step of a Goblin accumulation must be set to
be a public input of the circuit that performs the verification, so that
the verifier can extract that public input and use it as the commitment
to the previous table in the Merge verification.

For example, in ClientIVC the last Merge verification before the final
Goblin verification happens in the HidingKernel, so we need to add the
merged table commitments received by the Merge verifier inside the
HidingKernel to be public inputs of the HidingKernel.

After this PR, `MegaVerifier = UltraVerifier<MegaFlavor>` always expects
the inputs to be `PairingInputs` + commitments to ECC op tables. These
inputs are produced by the class `HidingKernelIO` (even though in the
future we might consider changing this name)

The PR required changes to various tests to accommodate the new
structure of the public inputs.

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: ledwards2225 <98505400+ledwards2225@users.noreply.github.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: ludamad <adam.domurad@gmail.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: notnotraju <raju@aztec-labs.com>
Co-authored-by: Lucas Xia <lucasxia01@gmail.com>
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>
Co-authored-by: Jean M <132435771+jeanmon@users.noreply.github.com>
Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Santiago Palladino <spalladino@users.noreply.github.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
Co-authored-by: ludamad <domuradical@gmail.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>
Co-authored-by: Sarkoxed <75146596+Sarkoxed@users.noreply.github.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

suyash67 committed:
chore: combined `uint` audit (#16030)

All `uint` audit-related changes. Note we are going to remove `uint`
module altogether, this PR/commit should serve as a fallback if we ever
need to use `uint` in future.

Commit:
Merge branch 'next' into merge-train/barretenberg

johnathan79717 committed:
chore: delete and ignore barretenberg/src/honk/keys/ (#16042)

These are auto-generated by bootstrap.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Introduce PublicInputComponent for native pub inputs; remove some appearances of PAIRING_POINT_SIZE; update HidingKernelIO

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Clean up PAIRING_POINTS_SIZE

federicobarbacovi commented on barretenberg/cpp/src/barretenberg/ecc/fields/field_declarations.hpp:
```diff
@@ -43,6 +43,9 @@ template <class Params_> struct alignas(32) field {
     using vec_out_buf = uint8_t**;
 
+    // The number of element required to represent field<Params_> in the public inputs of a circuit
+    static constexpr size_t PUBLIC_INPUTS_SIZE = Params::PUBLIC_INPUTS_SIZE;
```
Only needed because G2 uses the same interface as G1, so it needs to have a `PUBLIC_INPUTS_SIZE` constant

federicobarbacovi committed:
Clean up

federicobarbacovi committed:
Fix

federicobarbacovi committed:
fix: Fix tube proof construction (#16052)

Update the `prove_tube` function to take into account the new public
inputs of the Hiding kernel

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
feat: Link successive recursive Merge verifications (#16032)

We address https://github.com/AztecProtocol/barretenberg/issues/1351:
_Merge protocol table commitment consistency checks_.

More precisely, we ensure the Merge recursive verification at step $i+1$
uses as commitments to the previous state of the op queue the output of
the Merge recursive verification at step $i$. Instead of performing a
consistency check between the output at step $i$ and the input at step
$i+1$, we pass the output at step $i$ as the input at step $i+1$.

This PR also closes
https://github.com/AztecProtocol/barretenberg/issues/1473.

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/standardise_native_public_inputs

federicobarbacovi committed:
Remove appearances of IPA_CLAIM_SIZE

federicobarbacovi committed:
Remove appearances of PAIRING_POINTS_SIZE

federicobarbacovi committed:
Remove includes

federicobarbacovi committed:
Fixes

federicobarbacovi committed:
Reinstate tests

federicobarbacovi committed:
Clean up

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
chore: civc tests refactor (#16086)

In this PR:
* unify the logic of the two mock circuit producers for CIVC tests
* remove logic in CIVC to handle whether we are or are not receiving
precomputed vks and mocked vks, `accumulate` now expects a precomputed
verification key
* remove redundant tests

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/standardise_native_public_inputs

federicobarbacovi committed:
Post merge fixes

federicobarbacovi committed:
Fix after merge

federicobarbacovi committed:
Add constants to file

arielgabizon commented on barretenberg/cpp/src/barretenberg/commitment_schemes/pairing_points.hpp:
```diff
@@ -30,6 +29,8 @@ class PairingPoints {
     using VerifierCK = VerifierCommitmentKey<curve::BN254>;
 
   public:
+    static constexpr size_t PUBLIC_INPUTS_SIZE = 2 * Point::PUBLIC_INPUTS_SIZE;
```
As discussed, would be ideal to remove code duplication between native and in-circuit of PUBLIC_INPUT_SIZE

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

kashbrti committed:
chore: add a `PG_TAIL` proof type and queue type (#16156)

this PR adds a proof type and queue type for the tail kernel circuit so
we would be able to add the extra
operations required for this circuit in the client's IVC.

---------

Co-authored-by: AztecBot <tech@aztecprotocol.com>
Co-authored-by: Suyash Bagad <suyash@aztecprotocol.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: maramihali <mara@aztecprotocol.com>
Co-authored-by: Jonathan Hao <jonathan@aztec-labs.com>

Commit:
Merge branch 'next' into merge-train/barretenberg

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/standardise_native_public_inputs

federicobarbacovi committed:
Move public inputs constants to their own file

federicobarbacovi committed:
Define constants for sizes of special public inputs

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/standardise_native_public_inputs

federicobarbacovi committed:
Clean up

federicobarbacovi committed:
Merge remote-tracking branch 'origin/merge-train/barretenberg' into fb/standardise_native_public_inputs
